src/gt_pro.cpp

```diff
@@ -222,34 +235,32 @@ bool kmer_lookup_work(LmerRange* lmer_index, uint64_t* mmer_bloom, uint32_t* kme
 					// This kmer_tuple encodes the forward and reverse kmers at seq_buf[j...]
 					uint64_t kmer_tuple[2];
 					seq_encode<uint64_t, K>(kmer_tuple, seq_buf + j);
-
-					for (const auto& kmer: kmer_tuple) {
-						if ((mmer_bloom[(kmer & MAX_BLOOM) / 64] >> (kmer % 64)) & 1) {
-							const uint32_t lmer = kmer >> M2;
-							const auto range = lmer_index[lmer];
-							const auto start = range >> LEN_BITS;
-							const auto end = min(MAX_END, start + (range & MAX_LEN));
-							for (uint64_t z = start;  z < end;  ++z) {
-								const auto kmi = kmers_index[z];
-								const auto offset = kmi & 0x1f;
-								const auto snp_id = kmi >> 5;
-								const auto* snp_repr = snps + 3 * snp_id;
-								const auto low_bits = snp_repr[0] >> (62 - (offset * BITS_PER_BASE));
-								const auto high_bits = (snp_repr[1] << (offset * BITS_PER_BASE)) & BIT_MASK;
-								const auto db_kmer = high_bits | low_bits;
-								assert(lmer == (db_kmer >> M2));
-								if (kmer == db_kmer) {
-									// The set of kmers that covers this SNP within a given read cannot conflict.
-									// They will all map to the same virtual SNP.  Therefore, we can use the snp_id
-									// (which is really a virtual snp id) as a proxy for the real snp id.  Even
-									// though in general the mapping of virtual to real snp id is many to one,
-									// within each read it is always 1:1.
-									if (footprint.find(snp_id) == footprint.end()) {
-										kmer_matches.push_back(snp_id);
-										footprint.insert({snp_id, 1});
-									}
-								} else if (kmer < db_kmer) {
-									break;
+					const uint64_t kmer = min(kmer_tuple[0], kmer_tuple[1]);
+
+					if ((mmer_bloom[(kmer & MAX_BLOOM) / 64] >> (kmer % 64)) & 1) {
+						const uint32_t lmer = kmer >> M2;
+						const auto range = lmer_index[lmer];
+						const auto start = range >> LEN_BITS;
+						const auto end = min(MAX_END, start + (range & MAX_LEN));
+						for (uint64_t z = start;  z < end;  ++z) {
+							const auto kmi = kmers_index[z];
+							const auto offset = kmi & 0x1f;
+							const auto snp_id = kmi >> 5;
+							const auto* snp_repr = snps + 3 * snp_id;
+							const auto low_bits = snp_repr[0] >> (62 - (offset * BITS_PER_BASE));
+							const auto high_bits = (snp_repr[1] << (offset * BITS_PER_BASE)) & FULL_KMER;
+							const auto db_kmer = high_bits | low_bits;
+							// true but slow due to rc:
+							assert(lmer == (db_kmer >> M2) || lmer == (reverse_complement(db_kmer) >> M2));
+							if (kmer_tuple[0] == db_kmer || kmer_tuple[1] == db_kmer) {
+								// The set of kmers that cover this SNP within this given read won't conflict
+								// with each other, but may still belong to different virtual SNPs.  We need
+								// to identify the real SNP they all belong to, and increment the real SNP's
+								// counter just once for the read.
+								const auto snp = snp_repr[2] & SNP_MAX_REAL_ID;
```

Comment:
Note:  This is a minor bug fix.   The bug was introduced with virtual SNPs.  As the comment on the left says, I thought the fact that kmers come from he same read and therefore agree completely would imply that they would be encoded with the same SNP entry.  However, due to intervening kmers, they may be encoded with different virtual SNP ids.  So we can no longer keep the virtual SNP id as an identifier of the real SNP id.  Have to look up the real one.  Cost something, but not too much overall.